keps/sig-storage/2324-volume-snapshot-namespace-transfer/README.md

```diff
@@ -0,0 +1,538 @@
+- The VolumeSnapshot must be `readyToUse` before the transfer begins.
+- Any owners of the destination namespace must accept the transfer. By using the StorageTransfer API, this ensures an owner in the destination namespace is aware of the transfer.
+
+## Design Details
+
+This continues the discussion from [API for PVC Namespace Transfer](https://github.com/kubernetes/enhancements/pull/2110/) 
+in regards to the StorageTransfer API.
+
+A `StorageTransferRequest` resource will be created in the target namespace:
+
+```yaml
+apiVersion: storage.k8s.io/v1alpha1
+kind: StorageTransferRequest
+metadata:
+  name: kAnUOwpzso
+  namespace: target-namespace
+spec:
+  approvalName: xyji6OwynW
+  targetName: bar
+```
+
+`StorageTransferApproval` resource will be created in the source namespace:
+
+```yaml
+apiVersion: storage.k8s.io/v1alpha1
+kind: StorageTransferApproval
+metadata:
+  name: xyji6OwynW
+  namespace: source-namespace
+spec:
+  source:
+    name: foo
+    kind: VolumeSnapshot
+  secrets:
+    - name: secret-name
+      namespace: secret-namespace
+      type: secret-type # such as `csi.storage.k8s.io/provisioner-secret-name`
+  targetNamespace: target-namespace
+  requestName: kAnUOwpzso
+```
+
+NOTE: The name of the `StorageTransferApproval` and `StorageTransferRequest` should not be easy to guess,
+to prevent users from claiming resources inappropriately.
+
+ When matching `StorageTransferRequest` and `StorageTransferApproval` resources are detected, the 
+ transfer process is initiated. We have the following matching criteria for the `StorageTransferRequest` 
+ named `r` and the `StorageTransferApproval` `a`:
+ - `r.metadata.name == a.spec.requestName`
+ - `r.metadata.namespace == a.spec.targetNamespace`
+ - `r.spec.approvalName == a.metadata.name`
+
+Once matching is complete, a controller will begin the transfer process. For PVCs, this performs the following:
+- Adds a Finalizer to the PVC source-namespace\foo
+- Set the reclaim policy of the associated PersistentVolume Retain (if not already).
+- Delete the PVC source-namespace\foo.
+- Wait for all Pods to stop using the PVC source-namespace\foo
+- Bind The PersistentVolume to target-namespace\bar.
+- Create the PVC target-namespace\bar by copying the spec of source-namespace\foo and setting spec.volumeName appropriately.
+- Reset the PersistentVolume reclaim policy
+- Remove the Finalizer from the PVC source-namespace\foo
+
+For VolumeSnapshots, this performs the following:
+- Creates a VolumeSnapshot in the target namespace.
+- Binds the VolumeSnapshotContent to the newly created VolumeSnapshot. If this was a dynamically provisioned VolumeSnapshotContent, then it adjusts the VolumeSnapshotContent.Spec.Source to point to the newly created
+VolumeSnapshot.
+- Deletes the source VolumeSnapshot.
+
```

Comment:
"transfering" and relabeling snapshots has always been something that made me kinda nervous.  For a number of storage backends that use things like cow type snapshots or linked reference to versions this introduces a pretty big problem.  The problem is that you could now have a snapshot in ns: foo that's actually lined to a snapshot in ns: bar.  The two namespaces aren't aware of this linkage on the back end and neither can delete resources.

I know some have proposed that backends handle the reference counting in this case and do a sort of 'soft delete' but that doesn't seem great because "somebody" is paying for that storage that they *think* they deleted.

To transfer snapshots I don't have a great alternative proposal, other than create a PVC from a snapshot and transfer that instead.  I'm probably overly paranoid about this though and I don't think there are any other folks that have this concern.